homeassistant/components/media_player/hdmi_cec.py

```diff
@@ -26,9 +26,16 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     """Find and return HDMI devices as +switches."""
     if ATTR_NEW in discovery_info:
         _LOGGER.info("Setting up HDMI devices %s", discovery_info[ATTR_NEW])
-        add_entities(CecPlayerDevice(hass, hass.data.get(device),
-                                     hass.data.get(device).logical_address) for
-                     device in discovery_info[ATTR_NEW])
+        entities = []
+        for device in discovery_info[ATTR_NEW]:
+            hdmi_device = hass.data.get(device)
+            entities.append(CecPlayerDevice(
+                hass, hdmi_device, hdmi_device.logical_address,
+            ))
+        add_entities(entities)
```

Comment:
```py
add_entities(entities, True)
```

Reply:
https://github.com/home-assistant/home-assistant/blob/43676fcaf402994f18735bbd71a62fca54567594/homeassistant/helpers/entity_platform.py#L187-L193 tells that `update_before_add` should only be called inside tests.

Is this consistent with the API? [I think I repeat myself](https://github.com/home-assistant/home-assistant/pull/18753#discussion_r236829317)

Reply:
Also, because it updates those entities before adding (`update_before_add`), it returns

```
Nov 28 02:08:07 raspberrypi hass: 2018-11-28 02:08:07 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Nov 28 02:08:07 raspberrypi hass: Traceback (most recent call last):
Nov 28 02:08:07 raspberrypi hass:   File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
Nov 28 02:08:07 raspberrypi hass:     result = coro.send(None)
Nov 28 02:08:07 raspberrypi hass:   File "/home/homeassistant/venv/lib/python3.5/site-packages/homeassistant/helpers/entity_platform.py", line 343, in _async_add_entity
Nov 28 02:08:07 raspberrypi hass:     raise HomeAssistantError(msg)
Nov 28 02:08:07 raspberrypi hass: homeassistant.exceptions.HomeAssistantError: Entity id already exists: switch.hdmi_0
```

Reply:
The source you linked to is not the actual function that is used in `setup_platform`.

Here's the code that calls `setup_platform`:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_platform.py#L88-L91

Here's the sync version  of the function that is used to add entities in the call above:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_platform.py#L170

This will schedule the async version:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_platform.py#L179

Which will in turn schedule the final coroutine that does the work:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_platform.py#L199

So the sync version that you linked to is never called in production:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_platform.py#L187

Reply:
Regarding the race, your change to remove the state update from `_update` is correct. We should also add a separate callback method that calls `self.schedule_update_ha_state(True)`, and set this as callback for the cec network.

Reply:
Thanks for the explanation, I've missed async and sync versions when I was following the stack.